File: tax/opentax/scripts/fill_il.ts
// fill_il.ts — fill an Illinois return's key computed lines. IL PDFs use
// descriptive AcroForm field names ("Step 7: Line 51 - Enter your net income"),
// so fields are addressed by name. Usage:
//   deno run -A scripts/fill_il.ts <form> <template.pdf> <out.pdf> <netIncome> <tax>
// form: il1120st | il1120 | il1040
import { PDFDocument, StandardFonts } from "pdf-lib";

const [form, tmpl, outPath, netIncome, tax] = Deno.args;
const doc = await PDFDocument.load(await Deno.readFile(tmpl), { ignoreEncryption: true });
const f = doc.getForm();
// deno-lint-ignore no-explicit-any
const set = (name: string, t: string) => { try { (f as any).getTextField(name).setText(t); } catch { console.error("miss:", name); } };

const NET: Record<string, string> = {
  il1120st: "Step 7: Line 51 - Enter your net income",
  il1120: "Step 6: Line 50 - Enter your net income",
  il1040: "Net income. Subtract Line 12 from Line 11",
};
const TAX: Record<string, string> = {
  il1120st: "Step 8: Line 56 - Net replacement tax. Subtract Line 55 from Line 54",
  il1120: "Step 8: Line 61 - Net income and replacement tax",
  il1040: "Income tax. Multiply Line 13 by 4.95%",
};

if (+netIncome && NET[form]) set(NET[form], Math.round(+netIncome).toString());
if (TAX[form]) set(TAX[form], Math.round(+tax).toString());

const font = await doc.embedFont(StandardFonts.Helvetica);
f.updateFieldAppearances(font);
f.flatten();
await Deno.writeFile(outPath, await doc.save());
console.log(`filled ${form}: net ${netIncome}, tax ${tax} → ${outPath}`);

File: tax/opentax/scripts/fill_il.py
#!/usr/bin/env python3
"""Fill an Illinois return's key computed lines. IL PDFs use long descriptive
AcroForm field names; we match the target /Tx field by substring and set it via
pypdf, then flag NeedAppearances so viewers render the values."""
import sys
from pypdf import PdfReader, PdfWriter
from pypdf.generic import NameObject, BooleanObject

# form -> [(value, [substrings that identify the field, case-insensitive])]
SPECS = {
    "il1120st": [("net", ["line 51 - enter your net income"]),
                 ("tax", ["line 56 - net replacement tax"])],
    "il1120":   [("net", ["line 39 - net income"]),
                 ("tax", ["line 58 - total net income and replacement"])],
    "il1040":   [("net", ["illinois base income"]),
                 ("tax", ["total tax from page 1"])],
}


def find_key(flds, subs):
    for k, v in flds.items():
        if v.get("/FT") == "/Tx" and all(s in k.lower() for s in subs):
            return k
    return None


def fill(form, tmpl, out, net, tax):
    r = PdfReader(tmpl)
    flds = r.get_fields() or {}
    vals = {}
    for which, subs in SPECS[form]:
        amt = net if which == "net" else tax
        if amt == 0:
            continue
        key = find_key(flds, subs)
        if key:
            # Losses print in parentheses (IL convention); tax is never negative.
            vals[key] = f"({abs(round(amt)):,})" if amt < 0 else f"{round(amt):,}"
        else:
            print(f"  !! no field for {which} ({subs})")
    w = PdfWriter()
    w.append(r)
    for p in w.pages:
        try:
            w.update_page_form_field_values(p, vals)
        except Exception:
            pass
    # NeedAppearances so viewers render the set values
    try:
        w._root_object["/AcroForm"][NameObject("/NeedAppearances")] = BooleanObject(True)
    except Exception:
        pass
    with open(out, "wb") as fh:
        w.write(fh)
    # verify
    f2 = PdfReader(out).get_fields() or {}
    got = {k.split(" - ")[0][:16]: f2.get(k, {}).get("/V") for k in vals}
    print(f"  {form}: set {got}")


if __name__ == "__main__":
    form, tmpl, out, net, tax = sys.argv[1], sys.argv[2], sys.argv[3], float(sys.argv[4]), float(sys.argv[5])
    fill(form, tmpl, out, net, tax)
